### PR TITLE
For discussion: tools for processing certificate revocation

### PR DESCRIPTION
**Pull request by daviddrysdale**
OK, there's a bunch of stuff here, best reviewed commit-by-commit.  In fact, some of the commits might be a bad idea, so it's probably worth considering whether each commit is worthwhile before reviewing the detail of it.

At the bottom level, `asn1` gets one fix for optional values, plus some tweaks and some improved diagnostics.

Next level up is the `x509` library.  There are various small changes (typos, lint fixes, magic number removal), plus export things for external use, and refactoring a few bits of code for re-use.  This is mostly in preparation for adding proper support for working with CRLs, later in the commit chain.

On top of the CRL code I've also added a new non-fatal error processing mechanism that includes a lot more metadata.  This may not be a worthwhile idea, which is why it's a separate pair of commits that can be dropped fairly easily.  (OTOH if you think it's a good idea, I've got another branch squirrelled away somewhere that does something similar for errors in `Certificate` parsing...)

For both `asn1` & `x509`, this means we're going to be explicitly diverging from the upstream Go versions a lot more, which will make merges of upstream changes a bit harder.  If we don't want to do that, we can drop various of the commits.

Next up an aside: exporting more internal constants makes it possible to add `certcheck` and `crlcheck` tools that examine certs/CRLs and print output like the equivalent `openssl x509 -text` commands.

Finally, some more experimental ideas to add an `add-crl` entrypoint to the CT client code and the CTFE.  I probably wouldn't merge these yet, but it's interesting to see how a naive CRL log might work.

### CODE REVIEW

File: x509/revoked.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+	RevocationReason RevocationReasonCode
+	InvalidityDate   time.Time
+	Issuer           GeneralNames
+}
+
+func parseRevokedCertificate(pkixRevoked pkix.RevokedCertificate) (*RevokedCertificate, error) {
+	result := RevokedCertificate{RevokedCertificate: pkixRevoked}
+	for _, e := range pkixRevoked.Extensions {
+		if expectCritical, present := certExtCritical[e.Id.String()]; present {
+			if e.Critical && !expectCritical {
```
Comment (AlCutter) [inline x509/revoked.go]: Are these `e` the same type as the `e` up above at L184? Worth factoring this check out if so?

File: x509util/crlcheck/crlcheck.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%s: %v\n", filename, err)
+			errcount++
+			continue
+		}
+		if *verbose {
+			fmt.Print(x509util.CRLToString(certList))
+		}
+
+		if *caFile != "" {
```
Comment (AlCutter) [inline x509util/crlcheck/crlcheck.go]: Can the majority of this block be outside the `for`?

File: x509util/crlcheck/crlcheck.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"os"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509util"
+)
+
+var caFile = flag.String("ca", "", "CA certificate file")
+var verbose = flag.Bool("verbose", false, "Verbose output")
+
+func addCerts(filename string, pool *x509.CertPool) {
```
Comment (AlCutter) [inline x509util/crlcheck/crlcheck.go]: Is this used?

File: x509util/certcheck/certcheck.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509util"
+)
+
+var root = flag.String("root", "", "Root CA certificate file")
+var intermediate = flag.String("intermediate", "", "Intermediate CA certificate file")
```
Comment (AlCutter) [inline x509util/certcheck/certcheck.go]: There might be multiple of these, would a "chain" file in the format which Chrome etc. save them in be more helpful?

File: x509/names.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+func parseGeneralName(data []byte, gname *GeneralNames, withMask bool) ([]byte, error) {
+	var v asn1.RawValue
+	var rest []byte
+	var err error
+	rest, err = asn1.Unmarshal(data, &v)
+	if err != nil {
+		return nil, fmt.Errorf("x509: failed to unmarshal GeneralNames: %v", err)
+	}
+	switch v.Tag {
+	case 0:
```
Comment (AlCutter) [inline x509/names.go]: Add constants to name tags?

File: x509util/files.go [hunk 6]
```diff
@@ -52,12 +52,22 @@ func ReadPossiblePEMURL(target, blockname string) ([]byte, error) {
...
 	if strings.Contains(string(data), "BEGIN "+blockname) {
-		block, _ := pem.Decode([]byte(data))
-		if block != nil && block.Type == blockname {
-			return block.Bytes
+		rest := data
+		for {
+			var block *pem.Block
+			block, rest = pem.Decode(rest)
+			if block == nil {
+				break
```
Comment (AlCutter) [inline x509util/files.go]: Do you care if there's still stuff in `rest` here?

File: x509/errors.go [hunk 7]
```diff
@@ -5,13 +5,279 @@ import "fmt"
...
 
 	ErrMaxID
 )
 
 // idToError gives a template x509.Error for each defined ErrorID; where the Summary
 // field may hold format specifiers that take field parameters.
-var idToError = map[ErrorID]Error{}
+var idToError = map[ErrorID]Error{
+
+	ErrInvalidCertList: Error{
```
Comment (AlCutter) [inline x509/errors.go]: Just a thought - if you populated this map in a `func init()` from a list of Error structs you could avoid the possibility of a mismatched id -> Error mapping (i.e. you're currently relying on the thing to the left of the `:` being the same as the thing to the right of `ID:` in the struct).

**Comment by AlCutter** [comment]
I'm not sure about the changes to ct and ctfe (final 6 commits) - feels like those are experimental stuff which should go elsewhere - another personality etc. Would be best to lop those off into another PR, but the x509 and CRL stuff seems ok.

File: jsonclient/client.go [hunk 8]
```diff
@@ -32,6 +31,7 @@ import (
 	"time"
 
 	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/x509"
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply jsonclient/client.go]: Done.

File: x509/names.go (hunk 5 above)
Comment (daviddrysdale) [reply x509/names.go]: Done.

File: x509/revoked.go [hunk 9]
```diff
@@ -0,0 +1,358 @@
+// Copyright 2009 The Go Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

File: x509/revoked.go [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/certificate-transparency-go/asn1"
+	"github.com/google/certificate-transparency-go/x509/pkix"
+)
+
+// OID values for revocation extensions (TBSCertList.Extensions and RevokedCertificate.Extensions.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

File: x509/revoked.go [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+var (
+	OIDExtensionCRLNumber                = asn1.ObjectIdentifier{2, 5, 29, 20}
+	OIDExtensionDeltaCRLIndicator        = asn1.ObjectIdentifier{2, 5, 29, 27}
+	OIDExtensionIssuingDistributionPoint = asn1.ObjectIdentifier{2, 5, 29, 28}
+
+	OIDExtensionCRLReasons        = asn1.ObjectIdentifier{2, 5, 29, 21}
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: They're for different things -- I've split the comment to make clearer.

File: x509/revoked.go [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+// It has the same content as pkix.CertificateList, but the contents include parsed versions
+// of any extensions.
+type CertificateList struct {
+	Raw                asn1.RawContent
+	TBSCertList        TBSCertList
+	SignatureAlgorithm pkix.AlgorithmIdentifier
+	SignatureValue     asn1.BitString
+}
+
+// HasExpired reports whether now is past the expiry time of certList.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

File: x509/revoked.go [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+func (certList *CertificateList) HasExpired(now time.Time) bool {
+	return now.After(certList.TBSCertList.NextUpdate)
+}
+
+// Indication of whether extensions need to be critical or non-critical. Extensions that
+// can be either are omitted from the map.
+var listExtCritical = map[string]bool{
+	OIDExtensionAuthorityKeyId.String():           false,
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: They're in RFC section order; added comments to make that clear.

File: x509/revoked.go [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+	OIDExtensionIssuerAltName.String():            false,
+	OIDExtensionCRLNumber.String():                false,
+	OIDExtensionDeltaCRLIndicator.String():        true,
+	OIDExtensionIssuingDistributionPoint.String(): true,
+	OIDExtensionFreshestCRL.String():              false,
+	OIDExtensionAuthorityInfoAccess.String():      false,
+}
+
+var certExtCritical = map[string]bool{
+	OIDExtensionCRLReasons.String():        false,
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: As above.

File: x509/revoked.go [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+				return nil, fmt.Errorf("x509: extension %v marked critical, expect non-critical", e.Id)
+			} else if !e.Critical && expectCritical {
+				return nil, fmt.Errorf("x509: extension %v marked non-critical, expect critical", e.Id)
+			}
+		}
+		switch {
+		case e.Id.Equal(OIDExtensionAuthorityKeyId):
+			// RFC 5280 s5.2.1
+			var a authKeyId
+			if rest, err := asn1.Unmarshal(e.Value, &a); err != nil {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: I prefer to leave this as-is, as it's consistent with the style of lots of other instances across x509/

File: x509/revoked.go [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+			var aia []authorityInfoAccess
+			if rest, err := asn1.Unmarshal(e.Value, &aia); err != nil {
+				return nil, fmt.Errorf("x509: failed to unmarshal X.509 authority information: %v", err)
+			} else if len(rest) != 0 {
+				return nil, errors.New("x509: trailing data after X.509 authority information")
+			}
+
+			for _, v := range aia {
+				// GeneralName: uniformResourceIdentifier [6] IA5String
+				if v.Location.Tag != 6 {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

File: x509/revoked.go [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+					continue
+				}
+				if v.Method.Equal(oidAuthorityInfoAccessOcsp) {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509/revoked.go]: Done.

File: x509/revoked.go (hunk 1 above)
Comment (daviddrysdale) [reply x509/revoked.go]: They are both the same type (`Extension`) but the errors become distinct in a later commit...

File: x509util/crlcheck/crlcheck.go (hunk 3 above)
Comment (daviddrysdale) [reply x509util/crlcheck/crlcheck.go]: Yes, on lines 61-62.

Comment (daviddrysdale) [reply x509util/crlcheck/crlcheck.go]: Doh, no, that's the other file.  Removed.

File: x509util/crlcheck/crlcheck.go (hunk 2 above)
Comment (daviddrysdale) [reply x509util/crlcheck/crlcheck.go]: Done.

File: x509util/x509util.go [hunk 18]
```diff
@@ -33,113 +33,37 @@ import (
...
+		oid.Equal(x509.OIDExtensionSubjectDirectoryAttributes) ||
+		oid.Equal(x509.OIDExtensionInhibitAnyPolicy) ||
+		oid.Equal(x509.OIDExtensionPolicyConstraints) ||
+		oid.Equal(x509.OIDExtensionPolicyMappings) ||
+		oid.Equal(x509.OIDExtensionFreshestCRL) ||
+		oid.Equal(x509.OIDExtensionSubjectInfoAccess) ||
+		oid.Equal(x509.OIDExtensionAuthorityInfoAccess) ||
+		oid.Equal(x509.OIDExtensionCTPoison) ||
+		oid.Equal(x509.OIDExtensionCTSCT) {
 		return true
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509util/x509util.go]: Not sure I follow -- this is what `gofmt` gives.

File: x509util/x509util.go [hunk 19]
```diff
@@ -598,16 +522,16 @@ func CertificateToString(cert *x509.Certificate) string {
...
+		oid.Equal(x509.OIDExtensionKeyUsage) ||
+		oid.Equal(x509.OIDExtensionExtendedKeyUsage) ||
+		oid.Equal(x509.OIDExtensionAuthorityKeyId) ||
+		oid.Equal(x509.OIDExtensionBasicConstraints) ||
+		oid.Equal(x509.OIDExtensionSubjectAltName) ||
+		oid.Equal(x509.OIDExtensionCertificatePolicies) ||
+		oid.Equal(x509.OIDExtensionNameConstraints) ||
+		oid.Equal(x509.OIDExtensionCRLDistributionPoints) ||
+		oid.Equal(x509.OIDExtensionAuthorityInfoAccess) {
 		return true
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply x509util/x509util.go]: Ditto

File: x509util/files.go (hunk 6 above)
Comment (daviddrysdale) [reply x509util/files.go]: Yeah, this is in a loop that reads multiple PEM blocks from one source.

File: x509/errors.go (hunk 7 above)
Comment (daviddrysdale) [reply x509/errors.go]: Done.

File: x509util/x509util.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (AlCutter) [reply x509util/x509util.go]: Sorry, I think I got tab blindness by this stage.

File: x509util/x509util.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (AlCutter) [reply x509util/x509util.go]: same here

**Pull request merged**